compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,418 @@
+            });
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Char => {
+            // u4char as vendor extended type
+            let mut s = String::from("u4char");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Str => {
+            // u3str as vendor extended type
+            let mut s = String::from("u3str");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Never => {
+            // u5never as vendor extended type
+            let mut s = String::from("u5never");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Compound types
+        // () in Rust is equivalent to void return type in C
+        ty::Tuple(_) if ty.is_unit() => {
+            typeid.push('v');
+        }
+
+        // Sequence types
+        ty::Tuple(tys) => {
+            // u5tupleI<element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u5tupleI");
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, encode_c_types)?);
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // u5arrayI<array-length><element-type>E as vendor extended type
+            let mut s = String::from("u5arrayI");
+            s.push_str(&format!("{}", len.val().try_to_scalar()?.to_i64().ok()?));
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types)?);
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types)?);
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+        // respective builtin type encoding as specified by the Itanium C++ ABI if encode_c_types
+        // is true.
+        ty::Adt(adt_def, ..) => {
+            // <length><name>
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.opt_item_name(def_id)?);
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                s.push('v');
+            } else {
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
```

Comment:
The item name is not enough to uniquely identify a type. Different crates can have the same name but a different crate disambiguator. In addition you are missing the generic type parameters. I think you should just call into the type mangling code of the v0 symbol mangling scheme. At least for non-`#[repr(C)]` types.

Reply:
> Aren't types monomorphized at this time? Do you have an example and how I should handle it?

The second field of the Adt variant contains all substitutions.

Reply:
> Yes, this is part of a larger discussion of how to identify C char and integer type uses for cross-language LLVM CFI support, and the solution will also apply to C void when implemented.

That is unrelated. I am talking about how two crates with the same name may define a type with the same name but a different definition. You need to include the crate disambiguator to prevent accidentally interpreting one struct as another.

Reply:
> > Aren't types monomorphized at this time? Do you have an example and how I should handle it?
> 
> The second field of the Adt variant contains all substitutions.

If all types are monomorphized at this time, there won't be any substitutions, right?

> > Yes, this is part of a larger discussion of how to identify C char and integer type uses for cross-language LLVM CFI support, and the solution will also apply to C void when implemented.
> 
> That is unrelated. I am talking about how two crates with the same name may define a type with the same name but a different definition. You need to include the crate disambiguator to prevent accidentally interpreting one struct as another.

Is the crate disambiguator you're suggesting the one described in v0's RFC? Would just including the crate name along the item name help? What would happen with types that are re-exported in different crates but are expected to be interchangeable?

Reply:
> If all types are monomorphized at this time, there won't be any substitutions, right?

The AdtDef is identical for all generic instances of the struct/enum/union. It only contains the definition you write as user. Eg for the type `Foo<u8>` it may contain `struct Foo<T> { a: T }`. The subsitutions are what distinguishes `Foo<u8>` and `Foo<u16>`.

> Is the crate disambiguator you're suggesting the one described in v0's RFC?

Pretty much.

> Would just including the crate name along the item name help?

No, the crate name is not guaranteed to be unique. Only the combination of crate name and crate disambiguator.

> What would happen with types that are re-exported in different crates but are expected to be interchangeable?

The `DefId` is the same. A `DefId` represents the actual definition, not the path by which you write it. Re-exports are resolved by the name resolution code.

Reply:
In addition to crate & crate disambiguator, generating unique identifiers for nominal types also requires fully qualified paths to the type (or equivalent information encoded through other means, e.g., a hash like in a legacy mangling).

Reply:
> > If all types are monomorphized at this time, there won't be any substitutions, right?
> 
> The AdtDef is identical for all generic instances of the struct/enum/union. It only contains the definition you write as user. Eg for the type `Foo<u8>` it may contain `struct Foo<T> { a: T }`. The subsitutions are what distinguishes `Foo<u8>` and `Foo<u16>`.

I see. Thanks for pointing it out!

> 
> > Is the crate disambiguator you're suggesting the one described in v0's RFC?
> 
> Pretty much.
> 
> > Would just including the crate name along the item name help?
> 
> No, the crate name is not guaranteed to be unique. Only the combination of crate name and crate disambiguator.
> 
> > What would happen with types that are re-exported in different crates but are expected to be interchangeable?
> 
> The `DefId` is the same. A `DefId` represents the actual definition, not the path by which you write it. Re-exports are resolved by the name resolution code.

The LowerTypeTests pass just aggregates pointers by unique identifiers. It doesn't demangle or parse these ids, so it's unaware of name or type resolution (or DefIds), which means the encoding for re-exported types have to be exactly the same as the original type; otherwise, function pointers of functions that have parameters of the same type but with different paths won't be aggregated in the same group.

Reply:
> In addition to crate & crate disambiguator, generating unique identifiers for nominal types also requires fully qualified paths to the type (or equivalent information encoded through other means, e.g., a hash like in a legacy mangling).

Seem the comment above. That was the reason I didn't include the paths in the encoding.

Reply:
Given that, @bjorn3 and @tmiasko, what information do you think should be included in the encoding? (I think that including the crate disambiguator and the path would break re-exported types.)

Reply:
Adding the crate disambiguator and the def path won't cause re-exported types to be referenced using different symbols from different crates. For example:

```rust
//- crate1.rs
pub struct Foo;

//- crate2.rs
pub use crate1::Foo;

//-crate3.rs
fn foo(a: crate2::Foo) {}
```

will result `Foo` to be represented as something like `crate1[abcd]::Foo[0]` independent of the crate from which you reference it.

Reply:
Done. Please, take another look whenever you have a chance.

Reply:
> Adding the crate disambiguator and the def path won't cause re-exported types to be referenced using different symbols from different crates. For example:
> 
> ```rust
> //- crate1.rs
> pub struct Foo;
> 
> //- crate2.rs
> pub use crate1::Foo;
> 
> //-crate3.rs
> fn foo(a: crate2::Foo) {}
> ```
> 
> will result `Foo` to be represented as something like `crate1[abcd]::Foo[0]` independent of the crate from which you reference it.

It just occurred to me: is it the same for extern types?

Reply:
In general the original approach which included crate name and type name seems good. I think it would be fine to use it and add more information later.

What additional information to include? For nominal types (that includes extern types) I would include a full path and generic arguments if any. This should be sufficient to uniquely identify the type. The most straightforward implementation strategy seems to be a stable hash of those.

Reply:
I did some more testing (using 'e' for ForeignMod), and:

```
extern {
    type Type1;
    fn foo(_: &Type1);
}
```

is encoded as:

```
!1 = !{i64 0, !"_ZTSFvu3refIu89NvNeCsc2mK0tzeyH8_51sanitizer_cfi_emit_type_metadata_id_itanium_cxx_abi10{{extern}}5Type1EE"}
```

Which is a different encoding from what the actual extern type would be encoded to where it's defined, making them incompatible. So, I'll continue using the same encoding I'm currently using for `#[repr(C)]` user-defined types (i.e., <length><name>, with <name> being <unscoped-name>) for extern types (i.e., `ty::Foreign`), and `bug!` on ForeignMod, since it's not possible in `encode_ty_name()` because it's not used for `ty::Foreign`.

Reply:
Would you mind taking another look whenever you have a chance?

Reply:
I was expecting that to ensure compatibility we would later generalize the function signature type before encoding. For example, replace all pointer type parameters in the signature with a common type: `void*`. Does that sound right to you?

Reply:
I think by default we should provide as much comprehensible protection as possible, and also, for cross-language CFI support, be compatible with Clang's default (which does not generalize pointers). If needed, we can provide a generalize pointers options later, similarly to Clang's `-fsanitize-cfi-icall-generalize-pointers`.